File: linkedlist/intersect_linkedlist/get_intersect_node.go
package intersectlinkedlist

import "trouble_solver/linkedlist"

// 找出两个单链表相交的起始节点，如果两个链表没有交点，返回 null。
func getIntersectionNode(headA, headB *linkedlist.ListNode) *linkedlist.ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	AHelper, BHelper := headA, headB
	for AHelper != BHelper {
		if AHelper == nil {
			AHelper = headB
		} else {
			AHelper = AHelper.Next
		}
		if BHelper == nil {
			BHelper = headA
		} else {
			BHelper = BHelper.Next
		}
	}
	return AHelper
}

File: linkedlist/intersect_linkedlist/GetIntersectNode.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-10-07 10:41:55
 * @LastEditTime: 2023-01-09 11:24:45
 */
package linkedlist.intersect_linkedlist;

import linkedlist.ListNode;

public class GetIntersectNode {
    ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode helperA = headA, helperB = headB;
        while (helperA != helperB) {
            // if (helperA == null) {
            //     helperA = headB;
            // } else {
            //     helperA = helperA.next;
            // }

            // if (helperB == null) {
            //     helperB = headA;
            // } else {
            //     helperB = helperB.next;
            // }
            
            helperA = helperA == null ? headB : helperA.next;
            helperB = helperB == null ? headA : helperB.next;
        }
        return helperA;
    }
}
